File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-1.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} list1
     * @param {ListNode} list2
     * @return {ListNode}
     */
    mergeTwoLists(list1, list2) {
        let merged = new ListNode()
        let r = merged

        while (list1 && list2){
            if (list1.val > list2.val){
                merged.next = new ListNode(list2.val)
                list2 = list2.next
            }else if (list1.val < list2.val) {
                merged.next = new ListNode(list1.val)
                list1 = list1.next
            }else{
                merged.next = new ListNode(list1.val)
                merged = merged.next
                merged.next = new ListNode(list2.val)
                list1 = list1.next
                list2 = list2.next
            }
            merged = merged.next
            
        }

        if(list1){
            merged.next = list1
        }else{
            merged.next = list2
        }

        return r.next
        
    }
}

File: Data Structures & Algorithms/reorder-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
"""
save the nodes in an array bcos i need to know the tail and pretail
make pointer to the curr head for return
while ansHead.nxt & ansHead.nxt.nxt
save the ansHead.next in tmpHead(nxtHead)
pop out the curr tail and point the head.nxt to the tail
point the ansHead.nxt.nxt = tmpHead
peek at the nxt tail and point it to None
then ansHead to the tmpHead
"""
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        arr = []
        i = 0
        tmp1Head = head
        ansHead = head
        while tmp1Head:
            arr.append(tmp1Head)
            tmp1Head = tmp1Head.next
        
        # print([i.val for i in arr])
        while ansHead.next and ansHead.next.next:
            tHead = ansHead.next
            tail = arr.pop()
            ansHead.next = tail
            ansHead.next.next = tHead
            nxtTail = arr[-1]
            nxtTail.next = None
            ansHead = tHead